### PR TITLE
[SM-706] Extract Authorization From Create/Update Secret Commands

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to extract authorization checks from the create/update secret commands.

## Code changes

* **bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs:**
* **src/Core/SecretsManager/AuthorizationRequirements/SecretOperationRequirement.cs:**
* **bitwarden_license/src/Commercial.Core/SecretsManager/SecretsManagerCollectionExtensions.css:**
New secret authorization handler for refactoring authorization checks out of commands.

<br>

* **bitwarden_license/test/Commercial.Core.Test/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandlerTests.cs:** 
Unit tests for new secret authorization handler.

<br>

* **bitwarden_license/src/Commercial.Core/SecretsManager/Commands/Secrets/CreateSecretCommand.cs:**
* **src/Core/SecretsManager/Commands/Secrets/Interfaces/ICreateSecretCommand.cs:**
* **bitwarden_license/src/Commercial.Core/SecretsManager/Commands/Secrets/UpdateSecretCommand.cs:**
* **src/Core/SecretsManager/Commands/Secrets/Interfaces/IUpdateSecretCommand.cs:**
Extracting authorization from commands.

<br>

* **bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/Secrets/CreateSecretCommandTests.cs:**
* **bitwarden_license/test/Commercial.Core.Test/SecretsManager/Commands/Secrets/UpdateSecretCommandTests.cs:**
Updating command unit tests to remove authorization tests.

<br>

* **src/Api/SecretsManager/Controllers/SecretsController.cs:** 
Updating the secrets controller to use the new secret authorization handler.

* **src/Api/SecretsManager/Models/Request/SecretUpdateRequestModel.cs:** 
Updating the request models `ToSecret` method to include the organization ID. This is needed on the resource for authorization checks.

* **test/Api.IntegrationTest/SecretsManager/Controllers/SecretsControllerTests.cs:** 
Updating integration tests to actually test creating a secret with no project.

* **test/Api.Test/SecretsManager/Controllers/SecretsControllerTests.cs:** 
Updating unit tests to use the new authorization handler.

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Review by coltonhurst — APPROVED** [review]

File: test/Api.IntegrationTest/SecretsManager/Controllers/SecretsControllerTests.cs [hunk 1]
```diff
@@ -148,11 +148,8 @@ public async Task CreateWithoutProject_RunAsAdmin_Success()
         var (org, _) = await _organizationHelper.Initialize(true, true);
         await LoginAsync(_email);
 
-        var project = await _projectRepository.CreateAsync(new Project { OrganizationId = org.Id, Name = "123" });
```
Comment (Hinton) [inline test/Api.IntegrationTest/SecretsManager/Controllers/SecretsControllerTests.cs:151]: Can you create secrets without projects now? Seems to better align with the tests name now though 👍

File: src/Api/SecretsManager/Controllers/SecretsController.cs [hunk 2]
```diff
@@ -77,18 +81,19 @@ public async Task<SecretWithProjectsListResponseModel> ListByOrganizationAsync([
     [HttpPost("organizations/{organizationId}/secrets")]
     public async Task<SecretResponseModel> CreateAsync([FromRoute] Guid organizationId, [FromBody] SecretCreateRequestModel createRequest)
     {
-        if (!_currentContext.AccessSecretsManager(organizationId))
+        if (createRequest.ProjectIds != null && createRequest.ProjectIds.Length > 1)
         {
-            throw new NotFoundException();
+            throw new BadRequestException();
         }
```
Comment (Hinton) [inline src/Api/SecretsManager/Controllers/SecretsController.cs:87]: This seems better suited as a validation in the request model.

File: bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        Secret resource)
+    {
+        if (!_currentContext.AccessSecretsManager(resource.OrganizationId))
+        {
+            return;
+        }
+
+        switch (requirement)
+        {
+            case not null when requirement == SecretOperations.Create:
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs:38]: Can't you do something like this?

```suggestion
            case SecretOperations.Create:
```

File: bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        bool hasAccess;
+
+        switch (accessClient)
+        {
+            case AccessClientType.NoAccessCheck:
+                hasAccess = true;
+                break;
+            case AccessClientType.User:
+                var newProject = resource.Projects?.FirstOrDefault();
+                var access = (await _secretRepository.AccessToSecretAsync(resource.Id, userId, accessClient)).Write;
```
Comment (Hinton) [inline bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs:107]: We removed a null check here, is that fine?

**Review by Hinton — COMMENTED** [review]
Went through the authorization handler and seems to match the existing behavior. Nice refactor, some minor comments that might not be anything.

File: test/Api.IntegrationTest/SecretsManager/Controllers/SecretsControllerTests.cs (hunk 1 above)
Comment (Thomas-Avery) [reply test/Api.IntegrationTest/SecretsManager/Controllers/SecretsControllerTests.cs:151]: Admins can now.

File: bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs (hunk 4 above)
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs:107]: In the original code:

```
          case AccessClientType.User:
                var oldProject = secret.Projects?.FirstOrDefault();
                var newProject = updatedSecret.Projects?.FirstOrDefault();
                var accessToOld = oldProject != null && (await _projectRepository.AccessToProjectAsync(oldProject.Id, userId, accessClient)).Write;
                var accessToNew = newProject != null && (await _projectRepository.AccessToProjectAsync(newProject.Id, userId, accessClient)).Write;
                return accessToOld && accessToNew;
```

We check access to the `oldProject` which is really the current project before processing the update and then a possible new project assignment.

I changed the logic around to instead:

  Does the user have access to the current state secret? 

```
var access = (await _secretRepository.AccessToSecretAsync(resource.Id, userId, accessClient)).Write;
```

And then check if they have access to a possible new project assignment.

File: src/Api/SecretsManager/Controllers/SecretsController.cs (hunk 2 above)
Comment (Thomas-Avery) [reply src/Api/SecretsManager/Controllers/SecretsController.cs]: Resolved in https://github.com/bitwarden/server/pull/2896/commits/54ac00026b19b8f95f894f7501e7f2cdcac68aa7

File: bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    protected override async Task HandleRequirementAsync(AuthorizationHandlerContext context,
+        SecretOperationRequirement requirement,
+        Secret resource)
+    {
+        if (!_currentContext.AccessSecretsManager(resource.OrganizationId))
+        {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs:34]: I had tried a few different ways and settled on this. 

One was making a new authorization handler, but that makes the controller code messy, which we don't want.

```
        var secret = createRequest.ToSecret(organizationId);
        var authorizationResult = await _authorizationService.AuthorizeAsync(User, secret, SecretOperations.Create);
        var orgAccess =
            await _authorizationService.AuthorizeAsync(User, organizationId, SecretsManager.HasOrganizaitonAccess);
        if (!authorizationResult.Succeeded || !orgAccess.Succeeded)
        {
            throw new NotFoundException();
        }
```

Just making another requirement like

```
        var secret = createRequest.ToSecret(organizationId);
        var authorizationResult = await _authorizationService.AuthorizeAsync(User, secret, SecretOperations.Create, SecretOperations.HasSmAccess);
        if (!authorizationResult.Succeeded)
        {
            throw new NotFoundException();
        }
 ```
 
 Will require even more boilerplate of switch statements in every `EntityRequirementHanlder` to handle the new requirement case.
 
 I'm open to other suggestions if you know of another way?

File: src/Api/SecretsManager/Models/Request/SecretUpdateRequestModel.cs [hunk 6]
```diff
@@ -33,4 +33,14 @@ public Secret ToSecret(Guid id, Guid organizationId)
             Projects = ProjectIds != null && ProjectIds.Any() ? ProjectIds.Select(x => new Project() { Id = x }).ToList() : null,
         };
     }
+
+    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
+    {
+        if (ProjectIds != null && ProjectIds.Length > 1)
```
Comment (Hinton) [inline src/Api/SecretsManager/Models/Request/SecretUpdateRequestModel.cs:39]: You can use
```suggestion
        if (ProjectIds is { Length: > 1 })
```

File: src/Api/SecretsManager/Models/Request/SecretCreateRequestModel.cs [hunk 7]
```diff
@@ -32,4 +32,14 @@ public Secret ToSecret(Guid organizationId)
             Projects = ProjectIds != null && ProjectIds.Any() ? ProjectIds.Select(x => new Project() { Id = x }).ToList() : null,
         };
     }
+
+    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
+    {
+        if (ProjectIds != null && ProjectIds.Length > 1)
```
Comment (Hinton) [inline src/Api/SecretsManager/Models/Request/SecretCreateRequestModel.cs:38]:
```suggestion
        if (ProjectIds is { Length: > 1 })
```

File: bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Hinton) [reply bitwarden_license/src/Commercial.Core/SecretsManager/AuthorizationHandlers/Secrets/SecretAuthorizationHandler.cs:34]: In that case lets stick with the current approach.

File: src/Api/SecretsManager/Models/Request/SecretUpdateRequestModel.cs (hunk 6 above)
Comment (Thomas-Avery) [reply src/Api/SecretsManager/Models/Request/SecretUpdateRequestModel.cs]: resolved in https://github.com/bitwarden/server/pull/2896/commits/8834331a623f06b28bb6fe8f030482dad65a9c15

File: src/Api/SecretsManager/Models/Request/SecretCreateRequestModel.cs (hunk 7 above)
Comment (Thomas-Avery) [reply src/Api/SecretsManager/Models/Request/SecretCreateRequestModel.cs]: Resolved in https://github.com/bitwarden/server/pull/2896/commits/8834331a623f06b28bb6fe8f030482dad65a9c15

**Review by coltonhurst — APPROVED** [review]
Looks great, thank you!

**Review by Hinton — APPROVED** [review]

**Pull request merged**